File: bindings/aster_python/aster/codec.py
"""
aster.codec — Fory serialization codec.

Spec reference: §5.1–5.6 (serialization protocols), §5.3 (XLANG tags), §5.5 (ROW mode)

Wraps Apache Fory (pyfory) to provide three serialization modes:
- XLANG: Cross-language, tag-based type registration
- NATIVE: Python-native serialization (no tag requirement)
- ROW: Row-oriented format for random-access field reads

Also provides transparent zstd compression for payloads exceeding a
configurable threshold.
"""

from __future__ import annotations

import dataclasses
import typing
from dataclasses import dataclass, fields
from typing import Any, get_type_hints

import pyfory
import zstandard

from aster_python.aster.types import SerializationMode

# ── Default compression threshold ────────────────────────────────────────────

DEFAULT_COMPRESSION_THRESHOLD: int = 4096  # bytes


# ── @fory_tag decorator ──────────────────────────────────────────────────────


def fory_tag(tag: str):
    """Attach a Fory XLANG type tag to a dataclass.

    The *tag* string is split on the last ``/`` into
    ``(__fory_namespace__, __fory_typename__)``.  If there is no ``/``
    the namespace is the empty string.

    For XLANG mode, every user-defined type that appears in an RPC
    method signature **must** be decorated with ``@fory_tag``.
    """

    def decorator(cls):
        parts = tag.rsplit("/", 1)
        if len(parts) == 2:
            cls.__fory_namespace__ = parts[0]
            cls.__fory_typename__ = parts[1]
        else:
            cls.__fory_namespace__ = ""
            cls.__fory_typename__ = tag
        cls.__fory_tag__ = tag
        return cls

    return decorator


# ── Type-graph walking ───────────────────────────────────────────────────────

# Primitive types that don't need Fory registration.
_PRIMITIVES = frozenset({
    int, float, str, bool, bytes, bytearray, type(None),
})


def _is_primitive(tp: type) -> bool:
    """Return True if *tp* is a primitive that doesn't need registration."""
    return tp in _PRIMITIVES


def _unwrap_generic(tp: Any) -> list[Any]:
    """Extract inner types from generic aliases like list[X], dict[K,V], Optional[X]."""
    origin = getattr(tp, "__origin__", None)
    if origin is None:
        return []
    args = getattr(tp, "__args__", ()) or ()
    return list(args)


def _walk_type_graph(root_types: list[type]) -> list[type]:
    """Walk the type graph starting from *root_types* and return all
    dataclass types that need Fory registration (in dependency order,
    leaves first).

    Primitives and generic containers (list, dict, set, etc.) are
    skipped — only concrete dataclass types are collected.
    """
    visited: set[type] = set()
    result: list[type] = []

    def _visit(tp: Any) -> None:
        # Unwrap Optional, list, dict, etc.
        origin = getattr(tp, "__origin__", None)
        if origin is not None:
            for arg in _unwrap_generic(tp):
                _visit(arg)
            return

        if not isinstance(tp, type):
            return
        if _is_primitive(tp):
            return
        if tp in visited:
            return

        visited.add(tp)

        if not dataclasses.is_dataclass(tp):
            return

        # Visit fields first (dependency order)
        try:
            hints = get_type_hints(tp)
        except Exception:
            hints = {}

        for field in dataclasses.fields(tp):
            field_type = hints.get(field.name, field.type)
            _visit(field_type)

        result.append(tp)

    for t in root_types:
        _visit(t)

    return result


def _validate_xlang_tags(types: list[type]) -> None:
    """Raise ``TypeError`` if any dataclass type lacks a ``@fory_tag``.

    Only called for XLANG mode where tag-based registration is required.
    """
    for tp in types:
        if not dataclasses.is_dataclass(tp):
            continue
        if not hasattr(tp, "__fory_tag__"):
            raise TypeError(
                f"Type {tp.__qualname__} is used in XLANG mode but has no "
                f"@fory_tag decorator. All types must be tagged for XLANG "
                f"serialization."
            )


# ── Framework-internal types ─────────────────────────────────────────────────

# These are registered automatically by ForyCodec.
# Import here to avoid circular imports — the types themselves are defined
# in protocol.py which uses the fory_tag from this module (or its own copy).
_INTERNAL_TYPES: list[type] | None = None


def _get_internal_types() -> list[type]:
    """Lazily import and return framework-internal protocol types."""
    global _INTERNAL_TYPES
    if _INTERNAL_TYPES is None:
        from aster_python.aster.protocol import StreamHeader, CallHeader, RpcStatus
        _INTERNAL_TYPES = [StreamHeader, CallHeader, RpcStatus]
    return _INTERNAL_TYPES


# ── ForyCodec ────────────────────────────────────────────────────────────────


class ForyCodec:
    """Serialization codec wrapping Apache Fory (pyfory).

    Supports XLANG, NATIVE, and ROW serialization modes.

    Args:
        mode: The serialization mode to use.
        types: User-defined types that will be serialized/deserialized.
            For XLANG mode, all types must have ``@fory_tag`` decorators.
            For NATIVE mode, tags are optional.
        compression_threshold: Payloads larger than this (in bytes) are
            zstd-compressed.  Set to ``-1`` to disable compression.
    """

    def __init__(
        self,
        mode: SerializationMode = SerializationMode.XLANG,
        types: list[type] | None = None,
        compression_threshold: int = DEFAULT_COMPRESSION_THRESHOLD,
    ) -> None:
        self.mode = mode
        self.compression_threshold = compression_threshold
        self._cctx = zstandard.ZstdCompressor()
        self._dctx = zstandard.ZstdDecompressor()

        user_types = list(types) if types else []

        # Walk the type graph to discover nested dataclass types.
        all_types = _walk_type_graph(user_types)

        # For XLANG mode, validate that all types have tags.
        if mode == SerializationMode.XLANG:
            _validate_xlang_tags(all_types)

        # Add framework-internal types (always registered for XLANG).
        internal = _get_internal_types()
        for it in internal:
            if it not in all_types:
                all_types.append(it)

        self._registered_types = all_types

        # Create the Fory instance.
        if mode == SerializationMode.ROW:
            # ROW mode uses a separate Fory instance
            self._fory = pyfory.Fory()
        else:
            self._fory = pyfory.Fory()

        # Register all discovered types.
        for tp in all_types:
            tag = getattr(tp, "__fory_tag__", None)
            if tag is not None:
                ns = getattr(tp, "__fory_namespace__", "")
                tn = getattr(tp, "__fory_typename__", tp.__name__)
                self._fory.register_type(tp, namespace=ns, typename=tn)
            elif mode == SerializationMode.NATIVE:
                # NATIVE mode: register without tag
                self._fory.register_type(tp)
            # else: type is already validated above for XLANG

        # For ROW mode, build encoder/decoder if available
        self._row_encoder = None
        self._row_decoder = None
        if mode == SerializationMode.ROW:
            self._setup_row(user_types)

    def _setup_row(self, types: list[type]) -> None:
        """Set up ROW-mode encoder/decoder if pyfory supports it."""
        # pyfory ROW support check — may not be available in all versions
        if hasattr(pyfory, "create_row_encoder"):
            # This path is for future pyfory versions with ROW support
            pass
        # ROW mode falls back to standard serialization if not available

    def encode(self, obj: Any) -> bytes:
        """Serialize an object according to the configured mode.

        Returns the raw serialized bytes (without compression).
        Use :meth:`encode_compressed` if you want automatic compression.
        """
        data = self._fory.serialize(obj)
        return bytes(data)

    def decode(self, data: bytes, expected_type: type | None = None) -> Any:
        """Deserialize bytes into an object.

        Args:
            data: The serialized bytes.
            expected_type: Optional expected type for validation.

        Returns:
            The deserialized object.

        Raises:
            TypeError: If *expected_type* is given and the result doesn't match.
        """
        result = self._fory.deserialize(data)
        if expected_type is not None and not isinstance(result, expected_type):
            raise TypeError(
                f"Expected {expected_type.__qualname__}, "
                f"got {type(result).__qualname__}"
            )
        return result

    def encode_compressed(self, obj: Any) -> tuple[bytes, bool]:
        """Serialize and optionally compress.

        Returns:
            A ``(data, compressed)`` tuple where *compressed* is ``True``
            if zstd compression was applied.
        """
        raw = self.encode(obj)
        if (
            self.compression_threshold >= 0
            and len(raw) > self.compression_threshold
        ):
            return self._cctx.compress(raw), True
        return raw, False

    def decode_compressed(
        self, data: bytes, compressed: bool, expected_type: type | None = None
    ) -> Any:
        """Decompress (if needed) and deserialize.

        Args:
            data: The (possibly compressed) serialized bytes.
            compressed: Whether the data is zstd-compressed.
            expected_type: Optional expected type for validation.
        """
        if compressed:
            data = self._dctx.decompress(data)
        return self.decode(data, expected_type)

    def compress(self, data: bytes) -> bytes:
        """Compress raw bytes with zstd."""
        return self._cctx.compress(data)

    def decompress(self, data: bytes) -> bytes:
        """Decompress zstd-compressed bytes."""
        return self._dctx.decompress(data)

    def encode_row_schema(self) -> bytes:
        """For ROW mode: serialize the schema for hoisting.

        Returns the serialized schema bytes that should be sent as the
        first frame (with ROW_SCHEMA flag) so the receiver can decode
        subsequent ROW-encoded payloads.

        Raises:
            NotImplementedError: If ROW mode is not supported by the
                installed pyfory version.
        """
        if self.mode != SerializationMode.ROW:
            raise ValueError("encode_row_schema() is only valid in ROW mode")

        # Build a schema description from registered types
        schema_types = []
        for tp in self._registered_types:
            if not dataclasses.is_dataclass(tp):
                continue
            tag = getattr(tp, "__fory_tag__", tp.__qualname__)
            field_defs = []
            try:
                hints = get_type_hints(tp)
            except Exception:
                hints = {}
            for f in dataclasses.fields(tp):
                ft = hints.get(f.name, f.type)
                field_defs.append({
                    "name": f.name,
                    "type": _type_name(ft),
                })
            schema_types.append({
                "tag": tag,
                "fields": field_defs,
            })

        # Serialize the schema using standard Fory serialization
        return self.encode(schema_types)

    @property
    def registered_types(self) -> list[type]:
        """Return the list of types registered with this codec."""
        return list(self._registered_types)


def _type_name(tp: Any) -> str:
    """Return a human-readable name for a type."""
    origin = getattr(tp, "__origin__", None)
    if origin is not None:
        args = getattr(tp, "__args__", ())
        if args:
            arg_names = ", ".join(_type_name(a) for a in args)
            return f"{getattr(origin, '__name__', str(origin))}[{arg_names}]"
        return getattr(origin, "__name__", str(origin))
    if isinstance(tp, type):
        return tp.__name__
    return str(tp)

File: bindings/typescript/packages/aster/src/codec.ts
/**
 * Codec abstraction — serialization + optional zstd compression.
 *
 * Spec reference: S5.1 (serialization), S5.2 (compression)
 *
 * Provides:
 * - JsonCodec: JSON over UTF-8 bytes (testing, development)
 * - ForyCodec: @apache-fory/core XLANG (production cross-language)
 * - Zstd compression for payloads > threshold
 */

import { MAX_DECOMPRESSED_SIZE } from './limits.js';
import { WIRE_TYPE_KEY } from './decorators.js';
import { ContractViolationError } from './status.js';
import { getWireShape } from './generated.js';

/** Default compression threshold in bytes (4 KiB). */
export const DEFAULT_COMPRESSION_THRESHOLD = 4096;

/** Re-export WIRE_TYPE_KEY as wireType for API compatibility. */
export { WIRE_TYPE_KEY as wireType };

/** Get the wire type tag from a @WireType-decorated class. */
export function getWireType(cls: unknown): string | undefined {
  return (cls as any)?.[WIRE_TYPE_KEY];
}

/** Generic codec interface for serialization/deserialization. */
export interface Codec {
  encode(obj: unknown, hintType?: unknown): Uint8Array;
  decode(payload: Uint8Array, hintType?: unknown): unknown;
  encodeCompressed(obj: unknown, hintType?: unknown): [data: Uint8Array, compressed: boolean];
  decodeCompressed(payload: Uint8Array, compressed: boolean, hintType?: unknown): unknown;
}

// -- Zstd helpers (use node:zlib which has zstd since Node 21.7) --------------

let _zstdAvailable: boolean | undefined;
let _zstdCompress: ((data: Uint8Array) => Uint8Array) | undefined;
let _zstdDecompress: ((data: Uint8Array, maxSize: number) => Uint8Array) | undefined;

async function initZstd(): Promise<boolean> {
  if (_zstdAvailable !== undefined) return _zstdAvailable;
  try {
    const zlib = await import('node:zlib');
    if (typeof zlib.zstdCompressSync === 'function') {
      _zstdCompress = (data) => new Uint8Array(zlib.zstdCompressSync(Buffer.from(data)));
      _zstdDecompress = (data, maxSize) => {
        const result = zlib.zstdDecompressSync(Buffer.from(data), { maxOutputLength: maxSize });
        if (result.byteLength > maxSize) {
          throw new Error(`decompressed size ${result.byteLength} exceeds limit ${maxSize}`);
        }
        return new Uint8Array(result);
      };
      _zstdAvailable = true;
    } else {
      _zstdAvailable = false;
    }
  } catch {
    _zstdAvailable = false;
  }
  return _zstdAvailable;
}

// Try to init eagerly
initZstd().catch(() => {});

function zstdCompress(data: Uint8Array): Uint8Array | null {
  return _zstdCompress ? _zstdCompress(data) : null;
}

function zstdDecompress(data: Uint8Array): Uint8Array {
  if (!_zstdDecompress) {
    throw new Error('zstd decompression not available (requires Node 21.7+)');
  }
  return _zstdDecompress(data, MAX_DECOMPRESSED_SIZE);
}

// -- JsonCodec ----------------------------------------------------------------

/**
 * Simple JSON codec for cross-language interop and development.
 *
 * Strict mode: when ``decode`` is called with a ``hintType`` argument
 * (a @WireType-decorated class constructor), the codec validates that
 * every key in the decoded object matches a field declared on the
 * class. Unknown keys raise ``ContractViolationError`` -- the producer
 * owns the contract, and consumers must use the field names defined
 * by the producer's manifest. Validation walks nested objects and
 * arrays recursively, so a bad field at any depth fails loudly.
 *
 * If ``hintType`` is omitted (or is ``null``/``undefined``), decoding
 * is permissive and returns the raw parsed value -- the codec doesn't
 * know what shape to enforce. Callers that need strict validation
 * must always pass the expected type.
 */
export class JsonCodec implements Codec {
  private encoder = new TextEncoder();
  private decoder = new TextDecoder();
  private threshold: number;

  constructor(compressionThreshold = DEFAULT_COMPRESSION_THRESHOLD) {
    this.threshold = compressionThreshold;
  }

  encode(obj: unknown): Uint8Array {
    return this.encoder.encode(JSON.stringify(obj));
  }

  decode(payload: Uint8Array, hintType?: unknown): unknown {
    const parsed = JSON.parse(this.decoder.decode(payload));
    if (hintType && typeof hintType === 'function') {
      validateContractShape(parsed, hintType as new (...args: any[]) => any);
    }
    return parsed;
  }

  encodeCompressed(obj: unknown, _hintType?: unknown): [Uint8Array, boolean] {
    const data = this.encode(obj);
    if (data.byteLength > this.threshold && _zstdAvailable) {
      const compressed = zstdCompress(data);
      if (compressed && compressed.byteLength < data.byteLength) {
        return [compressed, true];
      }
    }
    return [data, false];
  }

  decodeCompressed(payload: Uint8Array, compressed: boolean, hintType?: unknown): unknown {
    if (compressed) {
      const decompressed = zstdDecompress(payload);
      return this.decode(decompressed, hintType);
    }
    return this.decode(payload, hintType);
  }
}

/**
 * Cached introspection result for a @WireType class.
 *
 * We cache the field name set + nested-class map per constructor so
 * the validator doesn't `new cls()` on every decode -- that would
 * re-run any side effects in the constructor (e.g. `id =
 * crypto.randomUUID()` initializers, allocator calls). The cache is
 * a WeakMap so it doesn't pin classes that are otherwise garbage.
 *
 * `null` for the cache value means "introspection failed once, don't
 * try again" (e.g. the constructor required positional args). The
 * validator will fall back to permissive decode for that class
 * forever.
 */
interface ClassShape {
  fieldNames: Set<string>;
  /** field name -> nested @WireType class to recurse into, if any. */
  nestedTypes: Map<string, new (...args: any[]) => any>;
  /** field name -> array element @WireType class, if any. */
  elementTypes: Map<string, new (...args: any[]) => any>;
}
const _shapeCache = new WeakMap<new (...args: any[]) => any, ClassShape | null>();

/**
 * Classes we've already warned about for the runtime-introspection
 * fallback. Keyed by constructor so we warn at most once per class
 * per process. This is the signal that aster-gen hasn't been run —
 * if it shows up in server logs, run `bunx aster-gen`.
 */
const _introspectFallbackWarned = new WeakSet<new (...args: any[]) => any>();

function introspectClass(
  cls: new (...args: any[]) => any,
): ClassShape | null {
  const cached = _shapeCache.get(cls);
  if (cached !== undefined) return cached;

  // Prefer the generated shape registry when available — it handles
  // empty arrays, nullable nested types, and non-default-constructible
  // classes (all of which break the runtime `new cls()` path below).
  // TODO(aster-gen): once every user runs `bunx aster-gen` and Fory JS
  // exposes a declarative typeInfo form, the runtime fallback can be
  // deleted entirely. Tracked in ffi_spec/ts-buildtime-audit.md.
  const generated = getWireShape(cls);
  if (generated) {
    const shape: ClassShape = {
      fieldNames: new Set(generated.fieldNameSet),
      nestedTypes: new Map(generated.nestedTypes),
      elementTypes: new Map(generated.elementTypes),
    };
    _shapeCache.set(cls, shape);
    return shape;
  }

  if (!_introspectFallbackWarned.has(cls)) {
    _introspectFallbackWarned.add(cls);
    const name = cls.name || '<anonymous>';
    console.warn(
      `[aster] ${name}: falling back to runtime introspection (new cls() + Object.keys). ` +
      `Run 'npx aster-gen' to get ` +
      `empty-array / nullable-nested / non-default-constructible support and faster decode validation.`,
    );
  }

  let template: any;
  try {
    template = new cls();
  } catch {
    // Class isn't default-constructible -- record a sentinel so we
    // never retry, and fall back to permissive decode forever.
    _shapeCache.set(cls, null);
    return null;
  }

  const fieldNames = new Set(Object.keys(template));
  const nestedTypes = new Map<string, new (...args: any[]) => any>();
  const elementTypes = new Map<string, new (...args: any[]) => any>();

  for (const [key, defaultValue] of Object.entries(template)) {
    if (defaultValue === null || defaultValue === undefined) continue;
    if (Array.isArray(defaultValue)) {
      // For arrays, sample the first element if any. Empty arrays
      // can't be introspected -- documented limitation.
      const sample = defaultValue[0];
      const elementCls = sample?.constructor as
        | (new (...args: any[]) => any)
        | undefined;
      if (
        elementCls &&
        elementCls !== Object &&
        typeof elementCls === 'function'
      ) {
        elementTypes.set(key, elementCls);
      }
      continue;
    }
    if (typeof defaultValue !== 'object') continue; // primitive (incl. enum members)
    const nestedCls = (defaultValue as object).constructor as
      | (new (...args: any[]) => any)
      | undefined;
    if (
      nestedCls &&
      nestedCls !== Object &&
      nestedCls !== Array &&
      nestedCls !== Date &&
      nestedCls !== Map &&
      nestedCls !== Set &&
      typeof nestedCls === 'function'
    ) {
      nestedTypes.set(key, nestedCls);
    }
  }

  const shape: ClassShape = { fieldNames, nestedTypes, elementTypes };
  _shapeCache.set(cls, shape);
  return shape;
}

/**
 * Strict shape validation: walks ``value`` against ``cls`` and throws
 * ``ContractViolationError`` if any object has keys not declared on
 * the corresponding @WireType class. Recurses into nested objects and
 * arrays so a bad field at any depth fails loudly with the dotted
 * path to the violation.
 *
 * Limitations (documented; tests pin them):
 *
 * - Nested types behind a `null` / `undefined` default are not
 *   recursed into. Top-level validation always runs.
 * - Empty array defaults can't be element-introspected. Top-level
 *   validation always runs.
 * - Date / Map / Set / typed-array fields are treated as opaque
 *   primitives -- their values may be objects on the wire but the
 *   validator doesn't try to recurse.
 * - Class generics are erased at runtime; the validator sees the
 *   default value of the generic field, not its declared type.
 */
function validateContractShape(
  value: unknown,
  cls: new (...args: any[]) => any,
  path = '',
): void {
  if (value === null || value === undefined) return;
  if (typeof value !== 'object' || Array.isArray(value)) return;

  const shape = introspectClass(cls);
  if (shape === null) return; // class isn't default-constructible

  const { fieldNames, nestedTypes, elementTypes } = shape;
  const dict = value as Record<string, unknown>;

  const unexpected: string[] = [];
  for (const key of Object.keys(dict)) {
    if (!fieldNames.has(key)) unexpected.push(key);
  }
  if (unexpected.length > 0) {
    const sanitized = sanitizeKeys(unexpected);
    const location = path || cls.name || 'unknown';
    const message =
      `contract violation at ${location}: unexpected JSON field(s) ` +
      `${JSON.stringify(sanitized)} (expected: ${JSON.stringify([...fieldNames].sort())})`;
    throw new ContractViolationError(message, {
      unexpected_fields: sanitized.join(','),
      location,
      expected_class: cls.name || 'unknown',
    });
  }

  // Recurse into nested @WireType objects + arrays using the cached
  // shape map. This avoids re-instantiating the class on every
  // decode (preserving constructor side effects) and is O(1) per
  // field after the first decode.
  for (const [key, child] of Object.entries(dict)) {
    if (child === null || child === undefined) continue;
    const nestedPath = path ? `${path}.${key}` : `${cls.name || 'value'}.${key}`;
    const nestedCls = nestedTypes.get(key);
    if (nestedCls && typeof child === 'object' && !Array.isArray(child)) {
      validateContractShape(child, nestedCls, nestedPath);
      continue;
    }
    const elementCls = elementTypes.get(key);
    if (elementCls && Array.isArray(child)) {
      for (let i = 0; i < child.length; i++) {
        const item = child[i];
        if (item && typeof item === 'object' && !Array.isArray(item)) {
          validateContractShape(item, elementCls, `${nestedPath}[${i}]`);
        }
      }
    }
  }
}

/**
 * Repr-quote unexpected key names for safe logging.
 *
 * Prevents log injection: keys can contain control chars, ANSI
 * escapes, newlines, or backslashes that would corrupt the error
 * message or terminal. We replace control chars with their escape
 * forms, cap each key's length, and cap the number of keys in the
 * list so a malicious client can't blow up log storage with
 * megabyte-long key names.
 */
function sanitizeKeys(keys: string[], maxCount = 5, maxLen = 80): string[] {
  const out: string[] = [];
  for (const k of keys.slice(0, maxCount)) {
    let s = String(k);
    if (s.length > maxLen) s = s.slice(0, maxLen) + '...(truncated)';
    // Escape control chars + non-printable bytes via JSON.stringify
    // (which produces a quoted string with backslash-escapes), then
    // strip the surrounding quotes to keep the inline form readable.
    const quoted = JSON.stringify(s);
    out.push(quoted.slice(1, -1));
  }
  if (keys.length > maxCount) {
    out.push(`...(+${keys.length - maxCount} more)`);
  }
  return out;
}

// -- Type graph walking -------------------------------------------------------

/** Root types that have already triggered the walkTypeGraph fallback warning. */
const _walkGraphFallbackWarned = new WeakSet<new (...args: any[]) => any>();

/**
 * Walk the type graph starting from root types, discovering nested @WireType
 * classes by inspecting default values of instances.
 *
 * Returns all discovered types in dependency order (leaves first), suitable
 * for registration with Fory. This is the TS equivalent of Python's
 * `_walk_type_graph()`.
 *
 * **Deprecation path.** Prefer the generated file from `bunx aster-gen`:
 * it walks the type graph at build time via the TS compiler API and emits
 * a topologically-ordered `WIRE_TYPES` list with full type information,
 * correctly handling empty arrays, optional fields, and non-default-
 * constructible classes. This runtime path warns once per root type when
 * invoked and will be removed after the Fory JS binding exposes a
 * declarative typeInfo schema. Tracked in
 * `ffi_spec/ts-buildtime-audit.md`.
 *
 * Limitations (compared to Python's dataclass introspection):
 * - Only discovers nested types whose default values are instances of @WireType classes
 * - Types in arrays, optionals, or maps that default to empty/null must be registered explicitly
 *
 * @param rootTypes - Classes decorated with @WireType
 * @returns All types in dependency order (leaves first)
 */
export function walkTypeGraph(rootTypes: (new (...args: any[]) => any)[]): (new (...args: any[]) => any)[] {
  for (const cls of rootTypes) {
    if (!_walkGraphFallbackWarned.has(cls)) {
      _walkGraphFallbackWarned.add(cls);
      const name = cls.name || '<anonymous>';
      console.warn(
        `[aster] walkTypeGraph(${name}): runtime type-graph reflection is in use. ` +
        `Run 'npx aster-gen' — the generated ` +
        `WIRE_TYPES list is built from AST types and handles cases this runtime path can't ` +
        `(empty arrays, nullable nested refs, non-default-constructible classes).`,
      );
    }
  }
  const visited = new Set<Function>();
  const ordered: (new (...args: any[]) => any)[] = [];

  function visit(cls: new (...args: any[]) => any): void {
    if (visited.has(cls)) return;
    visited.add(cls);

    // Check this class has a wire type tag
    const tag = (cls as any)[WIRE_TYPE_KEY];
    if (!tag) return;

    // Instantiate to discover fields and their default values
    try {
      const instance = new cls();
      for (const key of Object.keys(instance)) {
        const value = instance[key];
        if (value === null || value === undefined) continue;

        // Check if the value's constructor is a @WireType class
        const ctor = value?.constructor;
        if (ctor && ctor !== Object && ctor !== Array && ctor !== String &&
            ctor !== Number && ctor !== Boolean && (ctor as any)[WIRE_TYPE_KEY]) {
          visit(ctor as new (...args: any[]) => any);
        }

        // Check items in arrays for @WireType instances
        if (Array.isArray(value)) {
          for (const item of value) {
            const itemCtor = item?.constructor;
            if (itemCtor && (itemCtor as any)[WIRE_TYPE_KEY]) {
              visit(itemCtor as new (...args: any[]) => any);
            }
          }
        }
      }
    } catch {
      // If instantiation fails (e.g. required constructor args), skip walking fields
    }

    // Add after dependencies (leaves first)
    ordered.push(cls);
  }

  for (const cls of rootTypes) {
    visit(cls);
  }

  return ordered;
}

// -- ForyConfig ----------------------------------------------------------------

/**
 * Configuration for the Fory serializer.
 */
export class ForyConfig {
  /** Compression threshold in bytes (default: 4096). */
  compressionThreshold: number;
  /** Whether to use cross-language mode (xlang). Default: true. */
  xlang: boolean;

  constructor(opts?: { compressionThreshold?: number; xlang?: boolean }) {
    this.compressionThreshold = opts?.compressionThreshold ?? DEFAULT_COMPRESSION_THRESHOLD;
    this.xlang = opts?.xlang ?? true;
  }

  /** The resolved xlang mode (same as xlang field). */
  get resolvedXlang(): boolean {
    return this.xlang;
  }

  /** Convert to kwargs-style object for passing to Fory constructor. */
  toKwargs(): Record<string, unknown> {
    return {
      xlang: this.xlang,
      // Fory doesn't directly take compressionThreshold; pass via codec
    };
  }
}

/** Resolved ForyConfig with all defaults filled in. */
export interface ResolvedForyConfig {
  compressionThreshold: number;
  xlang: boolean;
  /** Resolved xlang (same as xlang). */
  resolvedXlang: boolean;
}

/**
 * Resolve a ForyConfig with defaults.
 */
export function resolveForyConfig(config?: ForyConfig | { compressionThreshold?: number; xlang?: boolean }): ResolvedForyConfig {
  const xlang = config?.xlang ?? true;
  return {
    compressionThreshold: config?.compressionThreshold ?? DEFAULT_COMPRESSION_THRESHOLD,
    xlang,
    resolvedXlang: xlang,
  };
}

// -- ForyCodec ----------------------------------------------------------------

/**
 * ForyCodec — cross-language serialization via @apache-fory/core.
 *
 * Wraps the Fory JS XLANG serializer for wire-compatible serialization
 * with Python's pyfory.
 *
 * @example
 * ```ts
 * import Fory from '@apache-fory/core';
 * const fory = new Fory({ compatible: true });
 * const codec = new ForyCodec(fory);
 * ```
 */
export class ForyCodec implements Codec {
  readonly fory: any;
  private threshold: number;
  private serializers = new Map<string, { serialize: any; deserialize: any }>();

  constructor(foryInstance: any, compressionThreshold = DEFAULT_COMPRESSION_THRESHOLD) {
    this.fory = foryInstance;
    this.threshold = compressionThreshold;
  }

  /** Register a type for serialization. */
  registerType(typeInfo: any): void {
    const { serialize, deserialize } = this.fory.registerSerializer(typeInfo);
    const name = typeInfo?.options?.typeName ?? typeInfo?.tag ?? String(typeInfo);
    this.serializers.set(name, { serialize, deserialize });
  }

  /**
   * Walk the type graph from root types and register all discovered
   * @WireType classes with Fory. Types are registered in dependency
   * order (leaves first).
   *
   * @param rootTypes - Classes decorated with @WireType
   * @param buildTypeInfo - Function that converts a class to a Fory typeInfo object.
   *   Receives (cls, wireTag) and should return the object to pass to registerType().
   */
  registerTypeGraph(
    rootTypes: (new (...args: any[]) => any)[],
    buildTypeInfo: (cls: new (...args: any[]) => any, wireTag: string) => any,
  ): void {
    const types = walkTypeGraph(rootTypes);
    for (const cls of types) {
      const tag = (cls as any)[WIRE_TYPE_KEY];
      if (tag && !this.serializers.has(tag)) {
        const typeInfo = buildTypeInfo(cls, tag);
        this.registerType(typeInfo);
      }
    }
  }

  encode(obj: unknown, hintType?: unknown): Uint8Array {
    // Convert plain object literals to typed instances when a hint type is provided.
    // This enables cross-language wire compatibility: plain { agent_id: 'x' } becomes
    // `new StatusRequest({ agent_id: 'x' })` so Fory can serialize it with the
    // correct wire tag instead of throwing "Failed to detect the Fory type".
    if (hintType && typeof hintType === 'function' && obj !== null && typeof obj === 'object' && !Array.isArray(obj)) {
      const ctor = hintType as new (...args: any[]) => any;
      obj = new ctor(obj);
    }
    const result = this.fory.serialize(obj);
    return new Uint8Array(result);
  }

  decode(payload: Uint8Array, _hintType?: unknown): unknown {
    return this.fory.deserialize(payload);
  }

  encodeCompressed(obj: unknown, hintType?: unknown): [Uint8Array, boolean] {
    const data = this.encode(obj, hintType);
    if (data.byteLength > this.threshold && _zstdAvailable) {
      const compressed = zstdCompress(data);
      if (compressed && compressed.byteLength < data.byteLength) {
        return [compressed, true];
      }
    }
    return [data, false];
  }

  decodeCompressed(payload: Uint8Array, compressed: boolean, hintType?: unknown): unknown {
    if (compressed) {
      const decompressed = zstdDecompress(payload);
      return this.decode(decompressed, hintType);
    }
    return this.decode(payload, hintType);
  }

  /**
   * Compress bytes using zstd (if available).
   * Returns null if zstd is unavailable or compression didn't help.
   */
  compress(data: Uint8Array): Uint8Array | null {
    if (!_zstdAvailable) return null;
    return zstdCompress(data);
  }

  /**
   * Decompress zstd-compressed bytes.
   */
  decompress(data: Uint8Array): Uint8Array {
    return zstdDecompress(data);
  }

  /**
   * Encode a row schema for a list of field names (for row-oriented data).
   * Returns a JSON-encoded schema descriptor.
   */
  encodeRowSchema(fields: string[]): Uint8Array {
    return new TextEncoder().encode(JSON.stringify({ fields }));
  }

  /**
   * Decode a row-oriented data payload using a previously decoded schema.
   * Reconstructs objects from parallel arrays.
   */
  decodeRowData(schemaBytes: Uint8Array, dataBytes: Uint8Array): unknown[] {
    const schema: { fields: string[] } = JSON.parse(new TextDecoder().decode(schemaBytes));
    const rows: unknown[] = this.decode(dataBytes) as unknown[];
    if (!Array.isArray(rows)) return [];
    return rows.map((row: unknown) => {
      if (!Array.isArray(row)) return row;
      const obj: Record<string, unknown> = {};
      for (let i = 0; i < schema.fields.length; i++) {
        obj[schema.fields[i]!] = row[i];
      }
      return obj;
    });
  }

  /**
   * Return the list of registered type names.
   */
  registeredTypes(): string[] {
    return [...this.serializers.keys()];
  }
}
